docs/src/components/home/ElementPointer.tsx

```diff
@@ -0,0 +1,113 @@
+import * as React from 'react';
+import Box, { BoxProps } from '@material-ui/core/Box';
+import Typography from '@material-ui/core/Typography';
+import { debounce } from '@material-ui/core/utils';
+
+const PointerContext = React.createContext<undefined | React.Dispatch<React.SetStateAction<Data>>>(
+  undefined,
+);
+
+export const withPointer = <T extends React.ElementType>(
+  Component: T,
+  options: { id: string; name: string },
+) => {
+  const WithPointer = (props: unknown) => {
+    const root = React.useRef<null | HTMLElement>(null);
+    const setData = React.useContext(PointerContext);
+    return (
+      <React.Fragment>
+        {/* @ts-ignore */}
+        <Component
+          ref={root}
+          {...props}
+          onMouseOver={(event: React.MouseEvent) => {
+            event.stopPropagation();
+            if (setData && root.current) {
+              setData({
+                id: options.id,
+                target: root.current,
+                name: options.name,
+              });
+            }
+          }}
+        />
+      </React.Fragment>
+    );
+  };
+
+  return WithPointer as T;
+};
+
+export type Data = { id: null | string; name: null | string; target: null | HTMLElement };
+
+export default function PointerContainer({
+  onElementChange,
+  ...props
+}: BoxProps & { onElementChange?: (data: Data) => void }) {
+  const container = React.useRef<null | HTMLDivElement>(null);
+  const [data, setData] = React.useState<Data>({
+    id: null,
+    name: null,
+    target: null,
+  });
+  const debouncedSetData = React.useRef(debounce(setData, 200));
+  React.useEffect(() => {
+    if (onElementChange) onElementChange(data);
+  }, [data, onElementChange]);
+  return (
+    <PointerContext.Provider value={debouncedSetData.current}>
```

Comment:
Reading a ref is not safe in concurrent rendering. Why do we need to debounce setting state in the first place?

Reply:
The original code is in this PR. I added info here https://github.com/mui-org/material-ui/pull/27663/files#r690947739